pkg/prober/prober.go

```diff
@@ -0,0 +1,150 @@
+package prober
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/improbable-eng/thanos/pkg/component"
+	"github.com/prometheus/common/route"
+)
+
+const (
+	healthyEndpointPath  = "/-/healthy"
+	readyEndpointPath    = "/-/ready"
+	okProbeFmt           = "thanos %v is %v"
+	errorProbeFmt        = "thanos %v is not %v. Reason: %v"
+	probeErrorHTTPStatus = 503
+	initialErrorFmt      = "thanos %s is initializing"
+)
+
+// Prober represents health and readiness status of given component.
+type Prober struct {
+	logger       log.Logger
+	componentMtx sync.Mutex
+	component    component.Component
+	readyMtx     sync.Mutex
+	readiness    error
+	healthyMtx   sync.Mutex
+	healthiness  error
+}
+
+// SetComponent sets component name of the Prober displayed in responses.
+func (p *Prober) SetComponent(component component.Component) {
+	p.componentMtx.Lock()
+	defer p.componentMtx.Unlock()
+	p.component = component
+}
+
+func (p *Prober) getComponent() component.Component {
+	p.componentMtx.Lock()
+	defer p.componentMtx.Unlock()
+	return p.component
+}
+
+// NewProber returns Prober representing readiness and healthiness of given component.
+func NewProber(component component.Component, logger log.Logger) *Prober {
+	initialErr := fmt.Errorf(initialErrorFmt, component)
+	prober := &Prober{
+		component:   component,
+		logger:      logger,
+		healthiness: initialErr,
+		readiness:   initialErr,
+	}
+	return prober
+}
+
+// RegisterInRouter registers readiness and liveness probes to mux.
+func (p *Prober) RegisterInRouter(router *route.Router) {
+	router.Get(healthyEndpointPath, p.probeHandlerFunc(p.IsHealthy, "healthy"))
+	router.Get(readyEndpointPath, p.probeHandlerFunc(p.IsReady, "ready"))
+}
+
+// RegisterInMux registers readiness and liveness probes to mux.
+func (p *Prober) RegisterInMux(mux *http.ServeMux) {
+	mux.HandleFunc(healthyEndpointPath, p.probeHandlerFunc(p.IsHealthy, "healthy"))
+	mux.HandleFunc(readyEndpointPath, p.probeHandlerFunc(p.IsReady, "ready"))
+}
+
+func (p *Prober) writeResponse(w http.ResponseWriter, probeFunc func() error, probeType string) {
+	err := probeFunc()
+	if err != nil {
+		http.Error(w, fmt.Sprintf(errorProbeFmt, p.getComponent(), probeType, err), probeErrorHTTPStatus)
+		return
+	}
+	if _, e := io.WriteString(w, fmt.Sprintf(okProbeFmt, p.getComponent(), probeType)); e == nil {
+		level.Error(p.logger).Log("msg", "failed to write probe response", "probe type", probeType, "err", err)
+	}
+}
+
+func (p *Prober) probeHandlerFunc(probeFunc func() error, probeType string) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, _ *http.Request) {
+		p.writeResponse(w, probeFunc, probeType)
+	}
+}
+
+// IsReady returns error if component is not ready and nil if it is.
+func (p *Prober) IsReady() error {
+	p.readyMtx.Lock()
```

Comment:
Let's Move to `sync.RWMutex` and do `Rlock` and `RUnlock` here (: There will be more readers than writerrs.

Reply:
Great, thanks I didn't know there is also RWMutex, TIL :)